src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Delete/Deleter.cs

```diff
@@ -0,0 +1,134 @@
+        private readonly Lazy<IConformanceProvider> _conformanceProvider;
+        private readonly IFhirDataStore _fhirDataStore;
+        private readonly ISearchService _searchService;
+        private readonly ResourceIdProvider _resourceIdProvider;
+
+        public Deleter(
+            IResourceWrapperFactory resourceWrapperFactory,
+            Lazy<IConformanceProvider> conformanceProvider,
+            IFhirDataStore fhirDataStore,
+            ISearchService searchService,
+            ResourceIdProvider resourceIdProvider)
+        {
+            _resourceWrapperFactory = EnsureArg.IsNotNull(resourceWrapperFactory, nameof(resourceWrapperFactory));
+            _conformanceProvider = EnsureArg.IsNotNull(conformanceProvider, nameof(conformanceProvider));
+            _fhirDataStore = EnsureArg.IsNotNull(fhirDataStore, nameof(fhirDataStore));
+            _searchService = EnsureArg.IsNotNull(searchService, nameof(searchService));
+            _resourceIdProvider = EnsureArg.IsNotNull(resourceIdProvider, nameof(resourceIdProvider));
+        }
+
+        public async Task<ResourceKey> DeleteAsync(DeleteResourceRequest request, CancellationToken cancellationToken)
+        {
+            var key = request.ResourceKey;
+
+            if (!string.IsNullOrEmpty(key.VersionId))
+            {
+                throw new MethodNotAllowedException(Core.Resources.DeleteVersionNotAllowed);
+            }
+
+            string version = null;
+
+            switch (request.DeleteOperation)
+            {
+                case DeleteOperation.SoftDelete:
+                    var emptyInstance = (Resource)Activator.CreateInstance(ModelInfo.GetTypeForFhirType(key.ResourceType));
+                    emptyInstance.Id = request.ResourceKey.Id;
+
+                    ResourceWrapper deletedWrapper = _resourceWrapperFactory.CreateResourceWrapper(emptyInstance, _resourceIdProvider, deleted: true, keepMeta: false);
+
+                    bool keepHistory = await _conformanceProvider.Value.CanKeepHistory(key.ResourceType, cancellationToken);
+
+                    UpsertOutcome result = await _fhirDataStore.UpsertAsync(new ResourceWrapperOperation(deletedWrapper, true, keepHistory, null, false, false, bundleOperationId: null), cancellationToken);
+
+                    version = result?.Wrapper.Version;
+                    break;
+                case DeleteOperation.HardDelete:
+                case DeleteOperation.PurgeHistory:
+                    await _fhirDataStore.HardDeleteAsync(key, request.DeleteOperation == DeleteOperation.PurgeHistory, cancellationToken);
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(request));
+            }
+
+            return new ResourceKey(key.ResourceType, key.Id, version);
+        }
+
+        public async Task<int> DeleteMultipleAsync(ConditionalDeleteResourceRequest request, CancellationToken cancellationToken)
```

Comment:
Thought: If we had `IAsyncEnumerable<ResourceKey>` we could pass them out as we delete them

Reply:
We could, but what benefit would that do us? We don't want to save the ids because it could be a large ammount of data.